src/emqx_os_mon.erl

```diff
@@ -189,19 +223,93 @@ ensure_check_timer(State = #{cpu_check_interval := Interval}) ->
         _ -> State#{timer := emqx_misc:start_timer(timer:seconds(Interval), check)}
     end.
 
-%% At startup, memsup starts first and checks for memory alarms,
-%% but emqx_alarm_handler is not yet used instead of alarm_handler,
-%% so alarm_handler is used directly for notification (normally emqx_alarm_handler should be used).
-%%The internal memsup will no longer trigger events that have been alerted,
-%% and there is no exported function to remove the alerted flag,
-%% so it can only be checked again at startup.
-ensure_system_memory_alarm(HW) ->
-    case erlang:whereis(memsup) of
-        undefined -> ok;
-        _Pid ->
-            {Total, Allocated, _Worst} = memsup:get_memory_data(),
-            case Total =/= 0 andalso Allocated/Total * 100 > HW of
-                true -> emqx_alarm:activate(high_system_memory_usage, #{high_watermark => HW});
-                false -> ok
-            end
+ensure_mem_check_timer(#{mem_check_timer := Ref} = State) when is_reference(Ref) ->
+    %% timer already started
+    State;
+ensure_mem_check_timer(State) ->
+    Interval = resolve_mem_check_interval(State),
+    case is_sysmem_check_supported() of
+        true ->
+            State#{mem_check_timer => emqx_misc:start_timer(timer:seconds(Interval), check_mem),
+                   mem_check_interval => Interval
+                  };
+        false ->
+            State
+    end.
+
+resolve_mem_check_interval(#{mem_check_interval := Seconds}) when is_integer(Seconds) ->
+    Seconds;
+resolve_mem_check_interval(_) ->
+    %% this only happens when hot-upgrade from older version (< 4.3.14, or < 4.4.4)
+    try
+        %% memsup has interval set API using minutes, but returns in milliseconds from get API
+        IntervalMs = memsup:get_check_interval(),
+        true = (IntervalMs > 1000),
+        IntervalMs div 1000
+    catch
+        _ : _ ->
+            %% this is the memsup default
+            60
+    end.
+
+is_sysmem_check_supported() ->
+    %% sorry Mac and Windows, for now
+    {unix, linux} =:= os:type().
+
+%% we still need to set memsup interval for process (not system) memory check
+do_set_mem_check_interval(Seconds) ->
+    Minutes = Seconds div 60,
+    _ = memsup:set_check_interval(Minutes),
+    ok.
+
+%% keep the time unit alignment with memsup, minmum interval is 60 seconds.
+do_resolve_mem_check_interval(Seconds) ->
+    case is_integer(Seconds) andalso Seconds >= 60 of
+        true -> Seconds;
+        false -> 60
+    end.
+
+resolve_sysmem_high_watermark(#{sysmem_high_watermark := SysHW}) -> SysHW;
+resolve_sysmem_high_watermark(_) ->
+    %% sysmem_high_watermark is not found in state map
+    %% get it from memsup
+    resolve_watermark(memsup:get_sysmem_high_watermark()).
+
+resolve_watermark(W) when W > 0 andalso W =< 1 ->
```

Comment:
Maybe check if it is a float here?  Otherwise, it might change what would be 1 % into 100 % (not that 1 % is quite realistic, but...)

Reply:
it'd be confusing if `1` and `1.0` differs ?

Reply:
Good point.  But `1` becomes ambiguous as well :thinking:

Reply:
removed this smart guess.